src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClientX/Handlers/TransportCreation/TransportCreationHandler.cs

```diff
@@ -0,0 +1,153 @@
+            {
+                throw new Exception("Hostname did not resolve");
+            }
+
+            // If there is an IP version preference, apply it
+            switch (request.IpAddressPreference)
+            {
+                case SqlConnectionIPAddressPreference.IPv4First:
+                    Array.Sort(ipAddresses, IpAddressVersionSorter.InstanceV4);
+                    break;
+
+                case SqlConnectionIPAddressPreference.IPv6First:
+                    Array.Sort(ipAddresses, IpAddressVersionSorter.InstanceV6);
+                    break;
+
+                case SqlConnectionIPAddressPreference.UsePlatformDefault:
+                default:
+                    // Not sorting necessary
+                    break;
+            }
+
+            // Attempt to connect to one of the matching IP addresses
+            // @TODO: Handle opening in parallel
+            Socket socket = null;
+            var socketExceptions = new List<Exception>();
+            foreach (var ipAddress in ipAddresses)
+            {
+                var ipEndpoint = new IPEndPoint(ipAddress, request.Port);
+                try
+                {
+                    socket = await OpenSocketAsync(ipEndpoint, ct).ConfigureAwait(false);
+                }
+                catch(Exception e)
+                {
+                    socketExceptions.Add(e);
+                }
+            }
+
+            // If no socket succeeded, throw
+            if (socket is null)
+            {
+                throw new AggregateException(socketExceptions);
+            }
+
+            // Create the stream for the socket
+            return new NetworkStream(socket);
+        }
+
+        private async Task<Socket> OpenSocketAsync(IPEndPoint ipEndpoint, CancellationToken ct)
+        {
+            var socket = new Socket(ipEndpoint.AddressFamily, SocketType.Stream, ProtocolType.Tcp) { Blocking = false, };
+
+            // Enable keep-alive
+            socket.SetSocketOption(SocketOptionLevel.Socket, SocketOptionName.KeepAlive, true);
+            socket.SetSocketOption(SocketOptionLevel.Tcp, SocketOptionName.TcpKeepAliveInterval, 1);
+            socket.SetSocketOption(SocketOptionLevel.Tcp, SocketOptionName.TcpKeepAliveTime, 30);
+
+            // Open the socket
+            try
+            {
+                await socket.ConnectAsync(ipEndpoint, ct).ConfigureAwait(false);
+            }
+            catch (SocketException e)
+            {
+                if (e.SocketErrorCode is not SocketError.WouldBlock)
+                {
+                    throw;
+                }
+            }
+
+            // Verify the socket is open
+            // @TODO: Configure timeout?
+            #if NET8_0_OR_GREATER
+            var pollTime = TimeSpan.FromSeconds(30);
+            #else
+            const int pollTime = 30000000; // 30s in us
+            #endif
+            if (!socket.Poll(pollTime, SelectMode.SelectWrite))
```

Comment:
Is socket.Poll not blocking? I wonder if this will cause this async method to block, before returning?

Reply:
Looks like this is only necessary in sync flow